### PR TITLE
GEODE-7090 Remove dependency on DataSerializer from membership classes

### PR DESCRIPTION
**Pull request by bschuchardt**
See https://cwiki.apache.org/confluence/display/GEODE/Move+membership+code+to+a+separate+gradle+sub-project

DataSerializableFixedID is used by most internal classes for
serialization.  Membership, in particular, was constrained in its
implementation to use only "fixed" serialization identifiers because our
other serialization mechanisms (DataSerialiable, PDX) are dependent on
membership and messaging.

This PR moves serialization of DataSerializableFixedID to a new
sub-project, geode-serialization in support of also separating
membership into its own sub-project.

Many files are touched by this PR.  The vast majority are due to a
change in the serialization interface that introduces a new parameter in
DataSerializableFixedID's toData and fromData methods.  This parameter
gives access to the serializer and to the Version of the
source/destination.  This parameter was added to begin moving from a
"static" based serialization service to an instance based service.

Serialization code concerning "DSFID"s has been moved from
InternalDataSerializer to the new class DSFIDSerialzerImpl.
InternalDataSerializer holds on to an instance of this class and
delegates DSFID serialization to it.

InternalDataSerializer also effects registration of Geode's DSFID
classes with this serializer.  Membership "Services" handles
registration of its own DSFID classes with the service, as do other
sub-projects like geode-web.

HeapDataOutputStream has been used by most of our serialization code to
provide a DataOutput for our toData methods.  Over the years this class
has accumulated a lot of adornments for PDX serialization that aren't
appropriate for a geode-serialization sub-project.  Consequently I've
moved much of its code into a superclass, BufferDataOutputStream, that
provides most of its functionality w/o needing to reference other geode-
core classes.

Version.java has been repackaged in this PR, moving it into a
"serialization" package.  This affected some rolling-upgrade and
backward-compatibility tests that were referring to
Version.CURRENT_ORDINAL.  These have been modified to get the "current
version ordinal" from VersionManager.

Finally, InternalDataSerializer injects a serializer into the new
sub-project that provideds an API for handling other types of
serializable objects like PDX.  This serializer is available in the
DataSerializableFixedID toData/fromData context parameter and is used by
membership services when serializing Geode DistributionMessages.

I don't think most of the files deserve scrutiny.  The primary files
affected are InternalDataSerializer and DSFIDFactory and the refactoring
of their code into DSFIDSerializer.  Aside from that,
does the new SerializationContext API make sense?  Could it be improved?

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by bschuchardt** [comment]
There are two unit tests that are failing on this branch.  One does nothing but assert that the cache server has no client "proxies" around.  The other is getting an NPE when a mock doesn't return a statistics object.  Neither test fails when run individually or under stress-testing.  I've run over a thousand runs in stress testing.  The first test is fairly new from Mario Ivanac.  The second was recently modified by Ryan.

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java [hunk 1]
```diff
@@ -484,6 +498,38 @@ public void testBigMessageIsFragmentedWhenBroadcast() throws Exception {
...
+    @Override
+    public int getDSFID() {
+      return DSFID;
+    }
+
+    public ByteHolder(byte[] payload) {
+      this.payload = payload;
+    }
+
+    public ByteHolder() {}
```
Comment (mcmellawatt) [inline geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java]: Default constructor isn't ever used (IntelliJ warning).

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java [hunk 2]
```diff
@@ -484,6 +498,38 @@ public void testBigMessageIsFragmentedWhenBroadcast() throws Exception {
...
+
+    @Override
+    public void toData(DataOutput out, SerializationContext context) throws IOException {
+      StaticSerialization.writeByteArray(payload, out);
+    }
+
+    @Override
+    public void fromData(DataInput in, DeserializationContext context)
```
Comment (mcmellawatt) [inline geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java]: ClassNotFoundException isn't ever thrown (IntelliJ warning).

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java [hunk 3]
```diff
@@ -484,6 +498,38 @@ public void testBigMessageIsFragmentedWhenBroadcast() throws Exception {
     doTestBigMessageIsFragmented(false, true);
   }
 
+  public static class ByteHolder implements DataSerializableFixedID {
```
Comment (mcmellawatt) [inline geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java]: Maybe it would make sense to move this class to the bottom of the test class so it isn't between test methods?  I don't feel super strongly about it, just an idea.

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java [hunk 4]
```diff
@@ -497,8 +543,11 @@ public void doTestBigMessageIsFragmented(boolean mcastEnabled, boolean broadcast
 
     // send a big message and expect fragmentation
     GMSMember recipient = broadcastMessage ? null : messenger.localAddress;
+    services.getSerializer().registerDSFID(ByteHolder.DSFID, ByteHolder.class);
     JoinRequestMessage msg = new JoinRequestMessage(recipient, sender,
-        new byte[(int) (services.getConfig().getDistributionConfig().getUdpFragmentSize() * (1.5))],
+        new ByteHolder(
+            new byte[(int) (services.getConfig().getDistributionConfig().getUdpFragmentSize()
+                * (1.2))]),
```
Comment (mcmellawatt) [inline geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java]: Just curious - why the change from 1.5 to 1.2?

File: geode-core/src/main/java/org/apache/geode/DataSerializer.java [hunk 5]
```diff
@@ -460,22 +460,13 @@ public static void writeInetAddress(InetAddress address, DataOutput out) throws
...
-    }
-
-    try {
-      InetAddress addr = InetAddress.getByAddress(address);
-      if (logger.isTraceEnabled(LogMarker.SERIALIZER_VERBOSE)) {
-        logger.trace(LogMarker.SERIALIZER_VERBOSE, "Read InetAddress {}", addr);
-      }
-      return addr;
-    } catch (UnknownHostException ex) {
-      throw new IOException("While reading an InetAddress", ex);
```
Comment (mcmellawatt) [inline geode-core/src/main/java/org/apache/geode/DataSerializer.java]: Was this exception handling superfluous?

File: geode-core/src/main/java/org/apache/geode/cache/AttributesFactory.java [hunk 6]
```diff
@@ -1490,10 +1489,10 @@ public static void validateAttributes(RegionAttributes<?, ?> attrs) {
   }
 
   private static class RegionAttributesImpl<K, V> extends UserSpecifiedRegionAttributes<K, V>
-      implements Cloneable, Serializable {
+      implements Cloneable {
     public Set<String> gatewaySenderIds;
     public Set<String> asyncEventQueueIds;
-    private static final long serialVersionUID = -3663000883567530374L;
+    // private static final long serialVersionUID = -3663000883567530374L;
```
Comment (mcmellawatt) [inline geode-core/src/main/java/org/apache/geode/cache/AttributesFactory.java]: Leftover commented out code

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ClientSideHandshakeImpl.java [hunk 7]
```diff
@@ -234,7 +234,8 @@ public ServerQueueStatus handshakeWithServer(Connection conn, ServerLocation loc
       // Read delta-propagation property value from server.
       // [sumedh] Static variable below? Client can connect to different
       // DSes with different values of this. It shoule be a member variable.
-      if (!communicationMode.isWAN() && currentClientVersion.compareTo(Version.GFE_61) >= 0) {
+      if (!communicationMode.isWAN() && currentClientVersion.compareTo(
```
Comment (mcmellawatt) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ClientSideHandshakeImpl.java]: I've seen a number of these changes where we've added a newline somewhere in the middle of an expression...not sure if you are doing these intentionally or your IntelliJ formatter has gone rogue.

File: boms/geode-all-bom/src/test/resources/expected-pom.xml [hunk 8]
```diff
@@ -871,10 +871,16 @@
...
+        <groupId>org.apache.geode</groupId>
+        <artifactId>geode-serialization</artifactId>
+        <version>1.11.0-SNAPSHOT</version>
+        <scope>compile</scope>
+      </dependency>
       <dependency>
         <groupId>org.apache.geode</groupId>
         <artifactId>geode-unsafe</artifactId>
-        <version>1.10.0-SNAPSHOT</version>
+        <version>1.11.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (bschuchardt) [reply boms/geode-all-bom/src/test/resources/expected-pom.xml]: That's what's in develop now.  In my branch it was put there by the update task.

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java (hunk 1 above)
Comment (bschuchardt) [reply geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java]: I'll address the warning - the default constructor is necessary for deserialization.  The constructor is registered with the serializer.
The test will fail if we remove that constructor.

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java (hunk 3 above)
Comment (bschuchardt) [reply geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java]: okay

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java (hunk 4 above)
Comment (bschuchardt) [reply geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java]: 1.5 was overkill with ByteHolder.  I think I made that change during debugging and just left it.

File: geode-core/src/main/java/org/apache/geode/DataSerializer.java (hunk 5 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/DataSerializer.java]: yes - not needed. We already handle this exception further up the call stack

File: geode-core/src/main/java/org/apache/geode/cache/AttributesFactory.java (hunk 6 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/cache/AttributesFactory.java]: oops

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ClientSideHandshakeImpl.java (hunk 7 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ClientSideHandshakeImpl.java]: I think that happened when I was trying to subclass Version to keep something in org.apache.geode.internal that could refer to all of its statics.  I gave up on that and undid the change but IJ seems to have left some formatting turds behind.

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java (hunk 2 above)
Comment (bschuchardt) [reply geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/membership/gms/messenger/JGroupsMessengerJUnitTest.java]: We probably have a lot of classes that IJ would give this warning about.

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java [hunk 9]
```diff
@@ -2246,7 +2255,7 @@ public void write(int b) throws IOException {
       try {
         ObjectOutput oos = new ObjectOutputStream(stream);
         if (stream instanceof VersionedDataStream) {
-          Version v = ((VersionedDataStream) stream).getVersion();
+          Version v = (Version) ((VersionedDataStream) stream).getVersion();
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: remove superfluous cast `(Version)`

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java [hunk 10]
```diff
@@ -1471,29 +1496,13 @@ public static void writeDSFID(DataSerializableFixedID o, DataOutput out) throws
...
-        String newClassName = o.getClass().getName();
-        String existingClassName = dsfidToClassMap.putIfAbsent(dsfid, newClassName);
-        if (existingClassName != null && !existingClassName.equals(newClassName)) {
-          logger.trace(LogMarker.SERIALIZER_WRITE_DSFID_VERBOSE,
-              "dsfid={} is used for class {} and class {}", dsfid, existingClassName, newClassName);
-        }
+    try {
+      if (dsfid != DataSerializableFixedID.NO_FIXED_ID) {
+        dsfidSerializer.writeDSFID(o, dsfid, out);
+        return;
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: Simplification! Bravo!

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java [hunk 11]
```diff
@@ -2348,6 +2349,10 @@ public static void invokeToData(Object ds, DataOutput out) throws IOException {
    */
   public static void invokeFromData(Object ds, DataInput in)
       throws IOException, ClassNotFoundException {
+    if (ds instanceof DataSerializableFixedID) {
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: Is there a reasonable way to eliminate this switch on type?

I see 5 instances of `instanceof DataSerializableFixedID` in our codebase.

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java [hunk 12]
```diff
@@ -2273,8 +2282,12 @@ public void write(int b) throws IOException {
    * @param out the output stream.
    */
   public static void invokeToData(Object ds, DataOutput out) throws IOException {
-    boolean isDSFID = ds instanceof DataSerializableFixedID;
     try {
+      boolean isDSFID = ds instanceof DataSerializableFixedID;
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: this variable is used only once—I suggest you inline the expr

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java [hunk 13]
```diff
@@ -2435,26 +2435,28 @@ public static Version getVersionForDataStream(DataInput in) {
...
-   * Get the {@link Version} of the peer or disk store that created this {@link DataInput}. Returns
+   * Get the {@link Version} of the peer or disk store that created this
+   * {@link DataInput}. Returns
    * null if the version is same as this member's.
    */
   public static Version getVersionForDataStreamOrNull(DataInput in) {
     // check if this is a versioned data input
     if (in instanceof VersionedDataStream) {
-      return ((VersionedDataStream) in).getVersion();
+      return (Version) ((VersionedDataStream) in).getVersion();
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: All the casts to `(Version)` in this file are superfluous. Recommend removal.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerSideHandshakeImpl.java [hunk 14]
```diff
@@ -159,7 +160,7 @@ public void handshakeWithClient(OutputStream out, InputStream in, byte endpointT
 
     Version v = Version.CURRENT;
     if (dos instanceof VersionedDataStream) {
-      v = ((VersionedDataStream) dos).getVersion();
+      v = (Version) ((VersionedDataStream) dos).getVersion();
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerSideHandshakeImpl.java]: cast to `(Version)` is superfluous—recommend removal

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java [hunk 15]
```diff
@@ -3615,16 +3593,18 @@ public int getDSFID() {
     }
 
     @Override
-    public void toData(DataOutput out) throws IOException {
-      super.toData(out);
+    public void toData(DataOutput out,
+        SerializationContext context) throws IOException {
+      super.toData(out, context);
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: why are we calling `super` here (and in `fromData()`? Both of those methods are no-ops

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java [hunk 16]
```diff
@@ -3615,16 +3593,18 @@ public int getDSFID() {
...
       out.writeInt(this.id);
       DataSerializer.writeObject(this.eventId, out);
     }
 
     @Override
-    public void fromData(DataInput in) throws IOException, ClassNotFoundException {
-      super.fromData(in);
+    public void fromData(DataInput in,
+        DeserializationContext context) throws IOException, ClassNotFoundException {
+      super.fromData(in, context);
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: see previous comment

File: geode-core/src/main/java/org/apache/geode/internal/DSFIDFactory.java [hunk 17]
```diff
@@ -1058,30 +1059,22 @@ private static InterestResultPolicyImpl readInterestResultPolicy(DataInput in)
   private static DataSerializableFixedID readDestroyOnDataStore(DataInput in)
       throws IOException, ClassNotFoundException {
     DataSerializableFixedID serializable = new DestroyRegionOnDataStoreMessage();
-    serializable.fromData(in);
+    serializable.fromData(in, null);
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/internal/DSFIDFactory.java]: These (3) calls to `fromData()` with context `null` concern be because I wonder what the recipient is going to do when context is `null`.

I see maybe 200 implementations of `DataSerializableFixedID`. All of those outside membership are still accessing the static data serializer (not referring to the context at all).

Is there a way to get ahold of a context here and pass it along, if only to prevent problems when we eventually go into the 200-odd implementations of DSFID and sever the ties to the static data serializer in favor of the context parameter?

File: geode-core/src/main/java/org/apache/geode/internal/DSFIDFactory.java [hunk 18]
```diff
@@ -423,9 +422,10 @@
  */
 public class DSFIDFactory implements DataSerializableFixedID {
 
-  private DSFIDFactory() {
-    // no instances allowed
-    throw new UnsupportedOperationException();
+  private final DSFIDSerializer serializer;
+
+  DSFIDFactory(DSFIDSerializer serializer) {
+    this.serializer = serializer;
```
[the remark this answers is not in the document]
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/DSFIDFactory.java]: good idea

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java (hunk 11 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: I'm open to suggestions on this and VersionedDataStream checks

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java (hunk 13 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: I think those casts must have crept in when I was trying to leave a subclass of Version in geode-core.  I'll remove them.

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java (hunk 15 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: I'm going to leave that alone - there are probably lots of instances where that's being done in geode-core.  Personally I would always invoke the superclass method like this is doing because I shouldn't assume that it's empty and will always stay empty.

File: geode-core/src/main/java/org/apache/geode/internal/DSFIDFactory.java (hunk 17 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/DSFIDFactory.java]: oh, good catch - I meant to deal with that before opening a PR

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java (hunk 12 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: okay, will do

File: geode-serialization/src/main/java/org/apache/geode/internal/serialization/AbstractSerializationContext.java [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+      // assume latest version
+      return Version.getCurrentVersion();
+    }
+  }
+
+  /**
+   * Get the Version of the peer or disk store that created this
+   * {@link DataOutput}.
+   */
+  Version getVersionForDataStream(DataOutput out) {
```
[the remark this answers is not in the document]
Comment (bschuchardt) [reply geode-serialization/src/main/java/org/apache/geode/internal/serialization/AbstractSerializationContext.java]: I've addressed the issue in AbstractSerializationContext but am leaving the InternalDataSerializer code alone.

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java [hunk 20]
```diff
@@ -287,6 +275,13 @@
    */
   @MakeNotStatic
   private static OldClientSupportService oldClientSupportService;
+
+  @MakeNotStatic
+  private static final DSFIDSerializer dsfidSerializer;
+
+  @MakeNotStatic
+  private static DSFIDFactory dsfidFactory;
```
[the remark this answers is not in the document]
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java [hunk 21]
```diff
@@ -339,7 +364,7 @@ public static String processOutgoingClassName(String name, DataOutput out) {
     }
     if (out instanceof VersionedDataStream) {
       VersionedDataStream vout = (VersionedDataStream) out;
-      Version version = vout.getVersion();
+      Version version = (Version) vout.getVersion();
```
[the remark this answers is not in the document]
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java (hunk 9 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java (hunk 12 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java (hunk 13 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/DSFIDFactory.java (hunk 17 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/DSFIDFactory.java]: done

File: geode-serialization/src/main/java/org/apache/geode/internal/serialization/AbstractSerializationContext.java [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+package org.apache.geode.internal.serialization;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+
+abstract class AbstractSerializationContext {
+
+  public Version getVersionForDataStream(DataInput in) {
+    // check if this is a versioned data input
+    if (in instanceof VersionedDataStream) {
```
[the remark this answers is not in the document]
Comment (upthewaterspout) [reply geode-serialization/src/main/java/org/apache/geode/internal/serialization/AbstractSerializationContext.java]: I had this thought too - does this API even need to support regular DataInput streams, or just versioned ones?
